Vetrov/statistics epic 3 3

Konstantin-vetro opened the pull request:
Эпик: Статистика
Выбранная архитектура: MVVM
Способ вёрстки: с помощью кода
Ссылка на доску: https://github.com/users/Skoch33/projects/2

Уважаемый ревьювер, была сложность по эпику только с очередями запросов. Добавил dispatchGroup и асинхронное выполнение задачи на каждую итерацию массива с ID NFT. Не знаю правильно ли я реализовал, готов к любым исправлениям по замечаниям) 
P.S. Все свои сервисы и запросы привел к готовым, реализованным шаблонам, имеющиеся  в проекте

Konstantin-vetro committed:
Добавил иконки корзины, лайка и звезды рейтинга

Konstantin-vetro committed:
Создал модель NFT

Konstantin-vetro committed:
создал сервис получения NFT по ID

Konstantin-vetro committed:
реализовал передачу массива индексом NFT на экран коллекции пользователя

Konstantin-vetro committed:
Создал вью модель NFT коллекции пользователя

Konstantin-vetro committed:
Сделал верстку экрана коллекции пользователя, добавил вью модель

Konstantin-vetro committed:
сделал верстку ячейки коллекции, реализовал нажатия на кнопки добавить в избранное и корзины

Konstantin-vetro committed:
Переименовал модель пользователя

Konstantin-vetro committed:
создал запросы для пользователей и по ID NFT

Konstantin-vetro committed:
произвел рефакторинг кода для сервисов пользователей и NFT, взяв стандартную реализацию

Konstantin-vetro committed:
переименовал старую модель на новую в ячейке таблицы

Konstantin-vetro committed:
переименовал модель и произвел рефакторинг кода по загрузке данных пользователей

Konstantin-vetro committed:
переименовал модель NFT

Konstantin-vetro committed:
переименовал модель NFT

Konstantin-vetro committed:
в ячейке коллекции переименовал модель

Konstantin-vetro committed:
переименовал модель и произвел рефакторинг кода по загрузке данных NFT

Konstantin-vetro committed:
убрал лишнее, переименовал изображения иконок, создал расширение для установки изображения иконок, чтобы код стал читабельнее

Konstantin-vetro committed:
добавил плейсхолдер картинке NFT, для темной темы обновил иконку

Konstantin-vetro committed:
убрал отмену задачи, иначе приложение вело себя некорректно

Konstantin-vetro committed:
добавил алерт при условии что загрузились не все NFT

Konstantin-vetro committed:
добавил обработку ошибки если NFT  c ID не загрузилось, добавил асинхронную очередь при работе с каждой task

LizaIlinykh commented on FakeNFT/Statistics/UserCard/UserCardViewController.swift:
```diff
@@ -163,9 +162,11 @@ final class UserCardViewController: UIViewController {
+            usersCollectionVC.title = "Коллекция NFT"
+            self?.navigationController?.pushViewController(usersCollectionVC, animated: true)
         }
     }
```
обьясню на примере onWebsiteButtonClick что не так во взаимдействии вью модели и вью

твоя вью сама вызывает замыкание для своего же собственного обновления. обновлять вью может только вью модель

вот как примерно должно выглядеть:

 ```
   @objc
    private func goToWebView() {
        /// вью просто сообщает вью модели о событии, не пытаясь обновиться самостоятельно
        viewModel.didTapOnWebsiteButton()
    }
```
    var onWebsiteButtonClick: ((URL) -> ())?
    
    func didTapOnWebsiteButton() {
        вью модель сама решает какую урлу передать вью и вызывает обновление вью
        guard let userURL = userWebsiteUrl else { return }
        onWebsiteButtonClick?(userURL)
    }
```
 private func bind() {
       вью получает любой урл который ему захочет отдать вью модель, а не пытается его вытянуть у вью модели и отображает его
        viewModel.onWebsiteButtonClick = { [weak self] url in
            let webView = WebViewViewController(webSite: url)
            self?.navigationController?.pushViewController(webView, animated: true)
        }
}
 
```

надеюсь понятно обьяснила в чем код не соответствует архитектурным правилам. если будут вопросы пиши. также часто делают роутер во вью модели , тогда ответственность за роутинг отделится от вью , что сделает код еще более чистым. советую поправить обработку события onCollectionButtonClick по примеру, который я тебе прислала. так получится лучше разобраться в MVVM

Konstantin-vetro replied:
Да, все понятно, исправил)
роутер не только в viper используется?

LizaIlinykh replied:
роутер это сущность которая отвечает за навигацию. еще есть такой паттерн как координатор, тоже занимается навигацией. эти паттерны можно юзать в любой архитектуре. важен сам факт распределения ответсвтенностей

LizaIlinykh commented on FakeNFT/Statistics/Rating/StatisticsViewModel.swift:
```diff
@@ -50,23 +50,18 @@ final class StatisticsViewModel: StatisticsViewModelProtocol {
                     self?.isDataLoading?(false)
+                case .failure(let error):
+                    self?.showError?(error)
+                    self?.isDataLoading?(false)
```
self?.isDataLoading?(false) повторяется два раза, можно вынести из свитча

Konstantin-vetro replied:
Вынес из свитча

LizaIlinykh commented on FakeNFT/Statistics/UsersCollection/UsersCollectionViewController.swift:
```diff
@@ -22,8 +54,90 @@ final class UserCollectionViewController: UIViewController {
+        let closeButton = AlertActionModel(
+            title: "Отмена",
+            style: .default
+        )
```
лучше все эти модели получать от вью модели

то есть ошибки могут быть самыми разными и тогда придется по каждую ошибку во вью делать разные методы для их отображения. хотя проще сделать один универсальный метод и просто в него передавать модель алерта который нужно отобразить

Konstantin-vetro replied:
Исправил

LizaIlinykh commented on FakeNFT/Statistics/UsersCollection/UsersCollectionViewController.swift:
```diff
@@ -4,14 +4,46 @@
+        return collectionView
+    }()
+    // MARK: - ViewModel
+    private var viewModel: UsersCollectionViewModelProtocol
```
можно сделать let

Konstantin-vetro replied:
Исправил

Konstantin-vetro committed:
исправил замечания ревьювера, а именно:

создал универсальный алерт для обработки сетевых ошибок
произвел рефакторинг кода обработки событий вью-модели пользователя
вью-модель в коллекции пользователя сделал константой
elf?.isDataLoading?(false) вынес из свитча во вью-модели статистики

Konstantin-vetro committed:
асинхронное выполнение замыкания перенес в сетевые сервисы

внес изменения во вью-модели статистики и пользователя

Konstantin-vetro committed:
Заново добавил кнопку сортировки

не отображалась по замечаниям ревьювера

Konstantin-vetro committed:
Реализовал получение тапнутого лайка NFT в коллекции пользователя

Создал модель
Создал запрос
Создал сервис профиля

Konstantin-vetro committed:
переименовал массив NFT пользователя

реализовал в ячейке коллекции установку состояния иконки лайка
во вью-модели реализовал сервис получения тапнутого лайка из профиля

Konstantin-vetro commented:
договорились в команде, что реализацию нажатия на корзинку или сердечко в коллекции НФТ через запрос на добавление/удаление НФТ в/из корзину/ы и в избранное/из избранного, будем делать перед мерджем в ветку develop
